src/Microsoft.OpenApi/Models/OpenApiComponents.cs

```diff
@@ -249,12 +250,205 @@ public void SerializeAsV3(IOpenApiWriter writer)
             writer.WriteEndObject();
         }
 
+        /// <summary>
+        /// Serialize <see cref="OpenApiComponents"/> to Open Api v3.0.
+        /// </summary>
+        public async Task SerializeAsV3Async(IOpenApiWriter writer)
+        {
+            if (writer == null)
+            {
+                throw Error.ArgumentNull(nameof(writer));
+            }
+
+            await writer.WriteStartObjectAsync();
+
+            // Serialize each referenceable object as full object without reference if the reference in the object points to itself.
+            // If the reference exists but points to other objects, the object is serialized to just that reference.
+
+            // schemas
+            await writer.WriteOptionalMapAsync(
+                OpenApiConstants.Schemas,
+                Schemas,
+                (w, key, component) =>
+                {
+                    if (component.Reference != null &&
+                        component.Reference.Type == ReferenceType.Schema &&
+                        component.Reference.Id == key)
+                    {
+                        component.SerializeAsV3WithoutReference(w);
+                    }
+                    else
+                    {
+                        component.SerializeAsV3(w);
+                    }
+                });
```

Comment:
The lambda should call `SerializeAsV3WithoutReferenceAsync` or `SerializeAsV3Async` and return a `Task`. Likewise in the rest of this method.